List Bot Tags On Site

HassanAbouelela opened the pull request:
Closes #687.

I implemented this feature in the most efficient way (in terms of network requests) possible. We download the entire bot project as a tarfile, extract all the tags from it, and cache them to the database.

Ideally, I'd also like to add a bit more flavor-info to the pages, such as when they were last updated, and the authors that have worked on them, but the only way to get that data from the API is to request data for a specific commit, which means making ~100 requests in total. We can't do that without being rate-limited, but it could be something to consider. I also considered downloading the git repo (as opposed to the content only), but it doesn't seem the API returns that, we'd need to clone with git.

Speaking of the 100 tags, I've gone through quite a few and they all look fine. They all should *render* fine, since the static deployment hits them all, but there could be some formatting oddities that I have missed. Feel free to tell me if so.

HassanAbouelela committed:
Add Setting For Static Builds

Adds an explicit setting for static builds instead of relying on the
environment variable.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add Tags To Content Listings

Adds bot tags to the content page, as well as a model to go along with
it.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add Tag Page Template

Add a template for the tag page itself, and add a route to use it.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add PyMDown Extensions

Adds the python-markdown extension pack as it offers better support for
things such as codeblocks.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add Content App To Navbar

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Export Tag Model As Top Level Model Object

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add Tag View Tests

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

ChrisLovering commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +39,111 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+            if len(content) > 100:
+                # Trim the preview to a maximum of 100 visible characters
+                # This causes some markdown to break, but we ignore that
+                content = content[:100] + "..."
```
```suggestion
                content = content[:100] + "…"
```
😡

HassanAbouelela replied:
Hahaha, I'm actually working on making the cropping here better, just keep your eyes peeled.

HassanAbouelela replied:
Stale due to f2ad3eed

ChrisLovering commented on pydis_site/apps/content/views/tags.py:
```diff
@@ -0,0 +1,56 @@
+
+        # Check for tags which can be hyperlinked
+        start = 0
+        while match := COMMAND_REGEX.search(content, start):
```
Would a `re.finditer` be more readable here? Rather than managing the iteration ourselves with `match.end()`.

HassanAbouelela replied:
Yes, that's what I originally did, but how do you do the substitution? Substituting by index would not work after the first match since the index would've moved, and substituting with something like re.sub wouldn't work either since the substitution still matches, so you'll end up on the first URL again.

ChrisLovering replied:
I made this suggestion without even looking at what this was doing. re.sub could be used here afaik, if you make the replacement a func rather than a string.
```py
def linkify_tag_ref(match: re.Match) -> str:
    link = reverse("content:tag", kwargs={"name": match.group("name")})
    return f"[{match.group()}]({link})"

...

content = COMMAND_REGEX.sub(linkify_tag_ref, content):
```

HassanAbouelela replied:
74226511

ChrisLovering commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +39,111 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+        tags = {}
+        for tag in get_tags():
+            content = frontmatter.parse(tag.body)[1]
+            if len(content) > 100:
```
I think we can get away with more than 100 chars in the preview, even on mobile, something ~300 is likely fine.
The problems arise when we start grabbing the first parts of a codeblock, and then we don't close them.
Is there a way for us to detect if we're "inside" a codeblock after trimming, and close it before the ellipses? 

Also, if the start of tags include a long hyperlink, the link length is counted towards this 100 character limit. This leads to some tags looking far shorter than other (example below). I wonder if there is anything we can do about this?

![image](https://user-images.githubusercontent.com/9753350/184507167-fe8cb056-680d-4adf-8543-bbf7d4fff85d.png)

HassanAbouelela replied:
It's not easy to fix this, at least in python, without effectively writing a full markdown interpreter. What I'm doing instead is re-writing this to do the splitting in javascript, where the text is already separated out in a convenient manner.

HassanAbouelela replied:
f2ad3eed

ChrisLovering reviewed (changes requested):
Just a couple a comments here.

Another observation is that this list is _very_ long. A paginated and/or searchable list could go a long way here.
I don't think this should be a blocker on this PR directly, but it should be something we look into in future.

Numerlor commented on pydis_site/apps/content/models/tag.py:
```diff
@@ -0,0 +1,17 @@
+        max_length=50,
+    )
+    body = models.TextField(help_text="The content of the tag.")
+    url = models.URLField(help_text="The URL to this tag on GitHub.")
```
does this need to be its own field instead of using the name and a configured github url?

HassanAbouelela replied:
It's a remnant from when I was still using the API and it'd return the URL. If we leave things as is, I'll drop it, but it might come in handy for the groups.

HassanAbouelela replied:
Ended up making it a property in b4911d0

Numerlor reviewed (changes requested):
It looks like this is currently missing proper support for tag groups (tags in directories under `tags/`), as there can be multiple tags with the same name under different dirs, and the github url template skips groups completely.

HassanAbouelela committed:
Set Link Color For Sub-elements

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Clean Up Tag Link Substitution

Co-authored-by: ChrisJL <ChrisLovering@users.noreply.github.com>
Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Enable Code Highlighting On All Content Pages

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Improve Tag Cropping

Move the tag cropping logic to the frontend, which makes it easier to
crop without crossing boundaries such as link or code block boundaries.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Move Tag URL To Property And Add Group

The URLs can be simply constructed using the other tag properties, so
they were removed from the database in favor of a property.

A group field was also added to support tags within groups.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add Tag Group Support

Adds support for tag groups in content. This involves some modification
to the routing, and templating.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Change Hyperlink Color On Hover

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Update Tests For Tag Groups

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

ChrisLovering commented on pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py:
```diff
@@ -0,0 +1,22 @@
+            name='group',
+            field=models.CharField(help_text='The group the tag belongs to.', max_length=50, null=True),
+        ),
+    ]
```
Do we really need a second migration file for un-merged changes?

Probably better to merge the two migrations together, so that there's only 1 migration to run after merge.

HassanAbouelela replied:
I figured it was easiest to write, and easiest to use for reviewers, but yeah we can merge them.

HassanAbouelela replied:
f2374900c4c83097c105b56de02ea82d66bd9466

HassanAbouelela committed:
Move GitHub strptime Format To Settings

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Unify Tag Migrations & Add Commit Model

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Add Tag Metadata

Uses the commit API to obtain tag metadata such as when it was last
edited, and by whom.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Better Split Up Tag Commit Messages

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Fix Tag Metadata For Static Builds

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Fix Tests For Tag Metadata

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

ChrisLovering commented on pydis_site/apps/content/models/tag.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class Commit(models.Model):
+    """A git commit."""
```
Possibly worth changing this to specify that it's a commit from the bot repo, since we're hard coding the URL

HassanAbouelela replied:
Yeah that's a fair point, c0384c62

ChrisLovering commented on pydis_site/apps/content/models/tag.py:
```diff
@@ -0,0 +1,75 @@
+    )
+    message = models.TextField(help_text="The commit message.")
+    date = models.DateTimeField(help_text="The date and time the commit was created.")
+    author = models.TextField(help_text="The person(s) who created the commit.")
```
Is it worth pluralising this to make it clear it's a list of authors, rather than a single author. From the format_users function, it seems that this is a serialised list of dicts? If so, probably worth mentioning that in the help text too, maybe even a comment describing the structure, or a link to the docs.

HassanAbouelela replied:
I called it author to follow the naming scheme of the API object, but it probably makes more sense to properly title it. In terms of the documentation, there's also the option of making the author another model, though it really only carries two important bits of info. Another option is to encode the author as a single string from the get-go if we wanted to avoid converting to and from JSON.

HassanAbouelela replied:
There's also the option of just not caring, it's unusual for the commit author to not be the committer as well, so it's not very important info to display.

ChrisLovering replied:
yea I don't think we need to go as far as another model, I was just surprised when reviewing that a field titled `author` was actually a serialized json structure _possibly_ containing multiple authors with additional info.

HassanAbouelela replied:
2343f7ab

ChrisLovering commented on pydis_site/apps/content/models/tag.py:
```diff
@@ -0,0 +1,75 @@
+            yield line
+
+    def format_users(self) -> collections.abc.Iterable[str]:
+        """Return a nice representation of the user(s)' name and email."""
```
Should this be format_authors, rather than format_users, so that we're not referring to it in 2 different ways

HassanAbouelela replied:
c0384c62

ChrisLovering commented on pydis_site/apps/content/models/tag.py:
```diff
@@ -0,0 +1,75 @@
+
+
+class Tag(models.Model):
+    """A tag from the python-discord server."""
```
Since we're hard coding this to the bot repo, should this instead be
```suggestion
    """A tag from the python-discord bot repository."""
```

ChrisLovering commented on pydis_site/apps/content/models/tag.py:
```diff
@@ -0,0 +1,75 @@
+    last_commit = models.ForeignKey(
+        Commit,
+        help_text="The commit this file was last touched in.",
+        null=True,
```
How can this be nullable?

HassanAbouelela replied:
The initial value can be null, since the data is only populated when requesting a single tag page (to avoid making 100 requests at once). Once the tag is accessed for the first time (since the last update), it's commit data is saved for future access.

HassanAbouelela replied:
Relevant code in the get_tag method:
https://github.com/python-discord/site/blob/92a42694b6ad1a29e5a21e0b3e57639528837113/pydis_site/apps/content/utils.py#L245-L248

ChrisLovering commented on pydis_site/apps/content/urls.py:
```diff
@@ -29,18 +29,39 @@ def __get_all_files(root: Path, folder: typing.Optional[Path] = None) -> list[st
         views.PageOrCategoryView.as_view(),
         name='page_category',
         distill_func=get_all_pages
-    ),
```
Is this diff needed?

HassanAbouelela replied:
It wasn't intentional, I'll see about removing it.

HassanAbouelela replied:
c0384c62

ChrisLovering commented on pydis_site/apps/content/utils.py:
```diff
@@ -1,22 +1,50 @@
+from .models import Commit, Tag
 
-def get_category(path: Path) -> Dict[str, str]:
+TAG_CACHE_TTL = datetime.timedelta(hours=1)
```
This is fine as is. However, what are your thoughts on using the TTLCache from [cachetools](https://pypi.org/project/cachetools/)? Just mentioning it was it's simply a deco around the function, as shown in the 3rd function [here](https://github.com/tkem/cachetools/#cachetools)

HassanAbouelela replied:
What would we apply the cache to here? The code checks the TTL against items from the database.

ChrisLovering replied:
Ah, this was my missunderstanding, I thought the ttl was used for an lru cache, but now I see what you mean. Ignore this.

HassanAbouelela replied:
Fair enough, thanks for the library suggestion though, it looks cool 😀

ChrisLovering commented on pydis_site/apps/content/views/tags.py:
```diff
@@ -0,0 +1,104 @@
+from pydis_site.apps.content import utils
+from pydis_site.apps.content.models import Tag
+
+COMMAND_REGEX = re.compile(r"`*!tags? (?P<name>[\w\d-]+)`*")
```
Tag groups are separated by spaces, rather than dashes. The below may work, but haven't tested if it causes any other oddities due to the space.
```suggestion
COMMAND_REGEX = re.compile(r"`*!tags? (?P<name>[\w\d ]+)`*")
```

HassanAbouelela replied:
Hmm, that dash is for tags, not tag groups, so I think it's still necessary (for instance `args-kwargs`). I hadn't considered tag groups at all when writing this, but I don't think they're separated by spaces either, I think the tag code does not support that. The relevant code for parsing tag commands is:

https://github.com/python-discord/bot/blob/dc9cb7910040f32d91aa546eec23b942a5a72203/bot/exts/info/tags.py#L66-L68


which leads me to believe that tags currently have to follow the `(tag-group)? tag-name` structure. I think it's simplest to write a regex which captures everything after `!tag` and handle the logic in the replacement function.

ChrisLovering replied:
Yea that seems reasonable, when i said separated by spaces I meant the group and the tag itself, I forgot tags could have hyphens in them too.

HassanAbouelela replied:
Actually it might be pretty difficult to accurately discern what's a tag group, and what's a tag with a normal sentence after it (for instance this tag link at the bottom of [args-kwargs](https://deploy-preview-763--pydis-static.netlify.app/pages/tags/args-kwargs/) which has a sentence after it). The only reliable way would be to either make a request to the page (reverse does not validate kwargs), or to use the get_tag function, but that can make an extra request (commit data) if the tag is not part of a group, neither of which are ideal solutions.

HassanAbouelela replied:
This regex achieves the first part of that:
```regex
`*!tags? (?P<first>[\w-]+)(?P<second> [\w-]+)?`*
```

It might optionally match a second word after the first one, and if it does, that could be a tag group combination

ChrisLovering replied:
Seems like it would do the job, obviously we'd need to update the tag model and the logic to get the url. So I'd be happy to merge this and have that as a separate PR if needed. Especially since we don't have any tag groups atm anyway.

HassanAbouelela replied:
I wrote the code for it, that was pretty simple, but I'm having unrelated trouble getting the test suite running. I'll probably have the fix out tomorrow

HassanAbouelela replied:
Fixed in efe784fe

ChrisLovering reviewed (changes requested):
A couple of comments, only the regex one is required, the rest are just observations.

Haven't reviewed js, css or html directly, but it looks good in the app itself.

HassanAbouelela committed:
Improve Tag Commit Naming

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Clarify Tag Model Docstring

Co-authored-by: ChrisJL <ChrisLovering@users.noreply.github.com>

HassanAbouelela committed:
Merge branch 'main' into bot-tags

HassanAbouelela committed:
Support Hyperlinked Tag Group Replacement

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Rename Tag Model Author Field

Renames the tag model `author` field to `authors` to better indicate
what's contained in the field, and updates the documentation.

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Merge branch 'main' into bot-tags

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

# Conflicts:
#	poetry.lock
#	pydis_site/settings.py
#	pyproject.toml

HassanAbouelela committed:
Merge branch 'main' into bot-tags

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

# Conflicts:
#	poetry.lock
#	pyproject.toml

jchristgit commented on pydis_site/apps/content/urls.py:
```diff
@@ -29,14 +29,35 @@ def __get_all_files(root: Path, folder: typing.Optional[Path] = None) -> list[st
 
+def get_all_tags() -> DISTILL_RETURN:
+    """Return all tag names and groups in static builds."""
+    groups = {None}
```
A comment couldn't hurt here as to why we start with `{None}`. I can understand it, as Chris has lovingly explained it to me.

HassanAbouelela replied:
3940e34b

jchristgit commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +55,258 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+    tar_file = client.get("/repos/python-discord/bot/tarball")
+    tar_file.raise_for_status()
+
+    client.close()
```
Why not use `with`?

HassanAbouelela replied:
e8b3745c

jchristgit commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +55,258 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+            # We still have an up-to-date commit entry
+            tag.last_commit = old_tag.last_commit
+
+        tag.save()
```
This part does not make sense to me. If the hash of the file changes, shouldn't we update it here? We don't seem to throw it away in that case either, but just stop updating.

HassanAbouelela replied:
We don't fetch the hashes here as documented in the approach methodology in the PR body since it would be too expensive in terms of ratelimits. There are ~100 tags, and it doesn't make sense to make an additional 100 requests to pre-load all the commits. When someone requests a tag, that commit data is fetched and saved, otherwise no commit data will be available until then.

jchristgit commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +55,258 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+    Commit.objects.filter(tag__isnull=True).delete()
+
+
+def get_tags() -> list[Tag]:
```
This typehint is wrong, because `return Tag.objects.all()` returns a `QuerySet`, **not** a `list[Tag]`. We could use `Iterable[Tag]` or possibly even `Container[Tag]`, but `list[Tag]` is wrong.

HassanAbouelela replied:
153059c

jchristgit commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +55,258 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+    """
+    path = path.split("/")
+    if len(path) == 2:
+        group, name = path[0], path[1]
```
```suggestion
        group, name = path
```

jchristgit commented on pydis_site/apps/content/utils.py:
```diff
@@ -27,8 +55,258 @@ def get_categories(path: Path) -> Dict[str, Dict]:
+        data.append(group)
+
+    # Sort the tags, and return them in the proper format
+    return {tag["title"]: tag for tag in sorted(data, key=lambda tag: tag["title"].lower())}
```
Use `str.casefold()` instead of `.lower()` for key functions.

```suggestion
    return {tag["title"]: tag for tag in sorted(data, key=lambda tag: tag["title"].casefold())}
```

jchristgit replied:
This may cause one of the 200 dictator linters to throw up due to line length.

HassanAbouelela replied:
426fc71b

jchristgit commented on pydis_site/apps/content/views/tags.py:
```diff
@@ -0,0 +1,124 @@
+# If the command does not include a group, the tag name will be in the `first` group
+# If there's a second word after the command, or if there's a tag group, extra logic
+# is necessary to determine whether it's a tag with a group, or a tag with text after it
+COMMAND_REGEX = re.compile(r"`*!tags? (?P<first>[\w-]+)(?P<second> [\w-]+)?`*")
```
This is a great comment, and I love it very much.

HassanAbouelela replied:
Thank you :D

jchristgit reviewed (changes requested):
Legend man. LEGEND!

I have small comments. If the comments are introspected, I am approving.

jchristgit reviewed (approved):
Great Behaviour From Joe Banks In The Dog Park Today.

HassanAbouelela committed:
Update pydis_site/apps/content/utils.py

Co-authored-by: Johannes Christ <jc@jchri.st>

HassanAbouelela committed:
Clean Up Typing

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Clarify Code Intentions

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Use `casefold` Instead Of `lower`

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Implicitly Close Client Using `with`

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

HassanAbouelela committed:
Merge branch 'main' into bot-tags

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

# Conflicts:
#	poetry.lock

ChrisLovering committed:
Merge branch 'main' into bot-tags
